### PR TITLE
Add a new method to identify the invoked API

### PR DESCRIPTION
**Pull request by tharikaGitHub**
## Purpose
This PR adds a new method to identify the invoked API from the collection of all APIs.

The initial change was https://github.com/wso2/wso2-synapse/pull/2083. But as this was a major refactoring, we have ported the changes from https://github.com/wso2/wso2-synapse/pull/2079 only.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 1]
```diff
@@ -158,6 +165,94 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+            if (isBound(r, synCtx) && r.canProcess(synCtx)) {
+                acceptableResources.add(r);
+            }
+        }
+        return acceptableResources;
+    }
+
+    public static API getSelectedAPI(MessageContext synCtx) {
+        API selectedApi;
+        API defaultAPI = null;
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:198]: defaultAPI is always null isn't it?

Comment (tharikaGitHub) [reply modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:198]: Yes that's correct default API is always null in this method. However this has been taken from the logic available in the AbstractApiHandler. https://github.com/wso2/wso2-synapse/blob/master/modules/core/src/main/java/org/apache/synapse/api/AbstractApiHandler.java#L47

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 2]
```diff
@@ -158,6 +165,89 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+        for (Resource r : resources.values()) {
+            if (isBound(r, synCtx) && r.canProcess(synCtx)) {
+                acceptableResources.add(r);
+            }
+        }
+        return acceptableResources;
+    }
+
+    public static API getSelectedAPI(MessageContext synCtx) {
+        API selectedApi;
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:197]: We can simply remove this variable.

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 3]
```diff
@@ -158,6 +165,89 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+        Set<Resource> acceptableResources = new LinkedHashSet<Resource>();
+        for (Resource r : resources.values()) {
+            if (isBound(r, synCtx) && r.canProcess(synCtx)) {
+                acceptableResources.add(r);
+            }
+        }
+        return acceptableResources;
+    }
+
+    public static API getSelectedAPI(MessageContext synCtx) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:196]: add javadoc comment on what we are doing with this utility method.

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 4]
```diff
@@ -158,6 +165,89 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+                    log.debug("Located specific API: " + api.getName() + " for processing message");
+                }
+                selectedApi = api;
+                return selectedApi;
+            }
+        }
+        return null;
+    }
+
+    public static boolean identifySelectedAPI(API api, MessageContext synCtx, List defaultStrategyApiSet) {
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:224]: do we have to make this public? what's the intension of this?

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java [hunk 5]
```diff
@@ -158,6 +165,94 @@ public static String getSubRequestPath(MessageContext synCtx) {
...
+            api.setLogSetterValue();
+            if (api.canProcess(synCtx)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Located specific API: " + api.getName() + " for processing message");
+                }
+                selectedApi = api;
+                return selectedApi;
+            }
+        }
+        if (defaultAPI != null && defaultAPI.canProcess(synCtx)) {
```
[the remark this answers is not in the document]
Comment (tharikaGitHub) [reply modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java]: no it is not getting executed. As discussed, will remove it.

File: modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java (hunk 4 above)
Comment (tharikaGitHub) [reply modules/core/src/main/java/org/apache/synapse/api/ApiUtils.java:221]: Yes we can make it private

**Review by isudana — APPROVED** [review]

**Pull request merged**